org.eclipse.m2e.core/src/org/eclipse/m2e/core/embedder/IMaven.java

```diff
@@ -148,11 +149,42 @@ MavenExecutionPlan calculateExecutionPlan(MavenProject project, List<String> goa
   MojoExecution setupMojoExecution(MavenProject project, MojoExecution execution, IProgressMonitor monitor)
       throws CoreException;
 
+  /**
+   * Instances should not be cached
+   *
+   * @since 2.1
+   */
+  public interface ConfigurationElement {
+    boolean exists();
+
+    // or firstChild() ? That's what it is actually but it does not look so nice when chained
+    ConfigurationParameter get(String name) throws NoSuchElementException;
+
+    // TODO: or getAll() or similar.
+    Stream<ConfigurationParameter> children(String name) throws NoSuchElementException;
```

Comment:
name is always unique except for collections (https://maven.apache.org/guides/mini/guide-configuring-plugins.html#mapping-collections-and-arrays). I don't think that we need to support that here, I one is interested in getting the collection she/he can call get("rootElementForCollection).as(Collection.class).

Reply:
That is a good point. But if one calls `get("rootElementForCollection).as(Collection.class)` the elements of the returned collection are (probably) of the type used in the Mojo, which are usually not available in a connector. I think it is better to have a collection/stream of ConfiguraitonElements here so that a caller can further query on this abstract/proxy level.

However, I have to admit I never actually had such case yet and therefore should create a test-case for that. Nevertheless I think it is relevant for the example given initially.